src/hotspot/share/opto/ifnode.cpp

```diff
@@ -628,31 +630,51 @@ const TypeInt* IfNode::filtered_int_type(PhaseGVN* gvn, Node* val, Node* if_proj
               return NULL;
             }
             case BoolTest::eq:
+              assert(!is_unsigned, "unsigned comparison is not supported");
               return cmp2_t;
             case BoolTest::lt:
-              lo = TypeInt::INT->_lo;
+              if (is_unsigned && lo >= 0) {
+                // cmp2 >= 0: val u<= cmp2 can only pass if val >= 0. Set val->_lo to 0.
+                lo = 0;
+              } else {
+                // The lower bound of val cannot be improved.
+                lo = TypeInt::INT->_lo;
+              }
               if (hi != min_jint) {
                 hi = hi - 1;
               }
               break;
```

Comment:
I think there's a bug here (case BoolTest::lt, unsigned comparison) . if cmp2 is [-1, 1] for instance, then lo and hi are unchanged. If the type of val is TypeInt::INT then the resulting type once join is applied is the interval [-1, 1]. But for instance 42 <u -1, 42 is in TypeInt::INT, -1 is in [-1, 1] but 42 is not in the resulting interval. I wonder if other cases in this method has similar issues. Can you fix that one and verify others?